pipeline-model-definition/src/main/resources/org/jenkinsci/plugins/pipeline/modeldefinition/ModelInterpreter.groovy

```diff
@@ -306,14 +306,20 @@ public class ModelInterpreter implements Serializable {
      * Executes a given closure in a "withEnv" block after installing the specified tools
      * @param agent The agent context we're running in
      * @param tools The tools configuration we're using
+     * @param root The root level configuration, if we're called within a stage. Can be null.
      * @param body The closure to execute
      * @return The return of the resulting executed closure
      */
-    def toolsBlock(Agent agent, Tools tools, Closure body) {
+    def toolsBlock(Agent agent, Tools tools, Root root = null, Closure body) {
+        def toolsList = []
+        if (tools != null) {
+            toolsList = tools.mergeToolEntries(root?.tools)
+        } else if (root?.tools != null) {
+            toolsList = root.tools.getToolEntries()
```

Comment:
Can't see if `getToolEntries()` is non null or not. But you seem to know that ;)